### PR TITLE
Add support for StringConcatFactory.makeConcatWithConstants (#9555)

### LINKED ISSUE #9555: [BUG] Scala 2.13 build with JDK 11 or 17 fails OpcodeSuite tests
From the Scala 2.13 PR https://github.com/NVIDIA/spark-rapids/pull/8592#discussion_r1367603174. There are some unresolved unit test failures in OpcodeSuite when building against JDK 11 or 17 (not with JDK 8).

```
 JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64 mvn -Pjdk11 test -pl udf-compiler -Dbuildver=330
- java lang string builder test - append *** FAILED ***
  OpcodeSuite.this.udfIsCompiled[T](ds1) was false (OpcodeSuite.scala:52)
- string test - + concat *** FAILED ***
  OpcodeSuite.this.udfIsCompiled[T](ds1) was false (OpcodeSuite.scala:52)
...
- throw a SparkException object *** FAILED ***
  OpcodeSuite.this.udfIsCompiled[T](ds1) was false (OpcodeSuite.scala:52)
- Conditional array buffer processing *** FAILED ***
  OpcodeSuite.this.udfIsCompiled[T](ds1) was false (OpcodeSuite.scala:52)

Tests: succeeded 161, failed 4, canceled 0, ignored 0, pending 0
```

The underlying issue seems to be a problem with `invokedynamic` which starts showing up in JDK 11 and 17 builds.  Should fix these so JDK 11 and JDK 17 builds can pass with Scala 2.13.

The code does run fine though with Java 11 and 17 runtime environments if built with JDK 8.

### PR DESCRIPTION
**Pull request by seanprime7**
This PR adds support for `StringConcatFactory.makeConcatWithConstants`.
4 of the existing unit tests started failing with JDK11 as the bytecode generation target due to the introduction of `StringConcatFactory.makeConcatWithConstants` in JDK9.
See #9555 for more details about the failures.

It resolves #9555

### CODE REVIEW

**Comment by seanprime7** [comment]
@abellina

**Review by abellina — COMMENTED** [review]
This looks good to me! It needs copyright updated, but the catalyst it produces matches what we did in JDK8/2.12.

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala [hunk 1]
```diff
@@ -78,6 +78,32 @@ class LambdaReflection private(private val classPool: ClassPool,
     }
   }
 
+  def lookupBootstrapMethod(constPoolIndex: Int): (CtMethod, Seq[Any]) = {
```
[the remark this answers is not in the document]
Comment (seanprime7) [reply udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala:81]: Done

**Comment by seanprime7** [comment]
> One thing separate from this PR we might consider is to try and simplify the concat catalyst expressions. With JDK8/scala 2.12 and with JDK11/scala 2.13 I see the same thing:
> 
> ```
> == Physical Plan ==
> AdaptiveSparkPlan isFinalPlan=false
> +- Project [x#2856, y#2857, z#2858, if (z#2858) concat(, , x#2856,  , y#2857, @@@,  true) else concat(, , y#2857,  , x#2856, !!!,  false) AS new#2862]
>    +- Exchange SinglePartition, REPARTITION_BY_NUM, [id=#883]
>       +- LocalTableScan [x#2856, y#2857, z#2858]
> ```
> 
> I think the concats cold be slightly simpler to remove that first empty string:
> 
> ```
> concat(x#2856, ' ', y#2857, '@@@',  ' true')
> ```
> 
> The code we are generating does the right thing, I am just saying it could be improved. @seanprime7 would this be an easy change? And it's ok for us to file an issue and do as a follow on, especially since it was already there before.

I've made a change not to generate `Concat` ops with empty strings in it.  If you want to further simplify the expression, you can add more simplification rules to `CatalystExpressionBuilder.scala`.

**Review by abellina — DISMISSED** [review]
The optimization (for empty space concat)  only went to scala 2.13, but that should be fine (the old code has been there for a long time).

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
@seanprime7 so I tried the optimization, but the catalyst plan looks the same?

```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [x#2856, y#2857, z#2858, if (z#2858) concat(, x#2856,  , y#2857, @@@ true) else concat(, y#2857,  , x#2856, !!! false) AS new#2862]
   +- Exchange SinglePartition, REPARTITION_BY_NUM, [id=#883]
      +- LocalTableScan [x#2856, y#2857, z#2858]
```

I'd rather split that off to a different issue, and back-out https://github.com/NVIDIA/spark-rapids/pull/10057/commits/047b41659cd48f8b4be1a332665efff46554a125, we can take a look at it later for both scala 2.12 and 2.13. 

Thoughts?

**Comment by seanprime7** [comment]
https://github.com/NVIDIA/spark-rapids/commit/047b41659cd48f8b4be1a332665efff46554a125 has been backed out

**Comment by abellina** [comment]
build

**Review by abellina — APPROVED** [review]

**Comment by abellina** [comment]
thanks @seanprime7

**Pull request merged**